server/src/subscribers/announcementSubscription.js

```diff
@@ -0,0 +1,72 @@
+const Queue = require('bull');
+const EmailServices = require('../services/EmailServices');
+const UserServices = require('../services/UserServices');
+
+const announcementSubscription = new Queue('newFrosh', {
+  redis: { port: process.env.REDIS_PORT, host: 'redis', password: process.env.REDIS_PASSWORD },
+});
+
+announcementSubscription.process((job, done) => {
+  console.log('Announcement Created!');
+  console.log(job.data);
+
+  if (job.data.unsubed === true) {
+    try {
+      const result = EmailServices.sendTemplateEmail(
+        {},
+        'unsubscribed',
+        [job.data.email],
+        'tech@orientation.skule.ca',
+      );
+
+      result.then((response) => {
+        console.log(response);
+      });
+
+      done();
+    } catch (error) {
+      done(error);
+    }
+  } else {
+    try {
+      const bulkEmailEntries = [];
+
+      UserServices.getAllUsers().then((users) => {
+        let count = 0;
+        let entries = [];
+        users.forEach((user) => {
+          if (user.canEmail === true) {
+            count++;
+            entries.push(user.email);
+
+            if (count === 20) {
+              bulkEmailEntries.push(entries);
+              count = 0;
+              entries = [];
+            }
+          }
+        });
+        if (entries.length > 0) {
+          bulkEmailEntries.push(entries);
+        }
+        const result = EmailServices.sendBulkTemplateEmail(
+          bulkEmailEntries,
+          'announcement',
+          { name: job.data.name, description: job.data.description },
+          'tech@orientation.skule.ca',
+        );
+
+        result.then((response) => {
+          console.log('response:');
+          console.log(response);
+        });
+
+        done();
```

Comment:
If you are going to do something like this, make sure that you only call `done()` in the `.then()` block of your Promise. In this code, `done()` will always be run, which is not correct as `done()` should only be called if the job succeeded (bull will retry the job later otherwise)

Reply:
In this code, if an error is thrown at some point in the Promise resolution, or if the Promise asynchronously throws an error, `done()` will have already been called, leading to the job potentially not being re-run